File: ruby/day2/grep.rb
#!/usr/bin/env ruby

# A simple grep


# Search for 'expr' in 'filename', if 'filename' is not provided use STDIN instead.
def grep(expr, filename='')
	if filename == ''
		lines = STDIN.readlines
	else
		begin
			File.open(filename, "r") do |f|
				lines = f.readlines
			end
		rescue SystemCallError => e
			STDERR.puts "#{filename}: Error: could not read from file (#{e})."
			return
		end
	end
	for line in lines
		if line =~ Regexp.new(expr)
			puts "#{filename}: #{line}"
		end
	end

end


if ARGV.size < 1
	# No arguments: print usage message
	STDERR.puts "Usage: #{File.basename($0)} expr file..."
elsif ARGV.size ==1
	# One argument: assume it's the search expression and search in STDIN
	grep(ARGV[0])
else
	# Multiple arguments: the first is the search term, the rest are the files to be searched
	expr = 	ARGV[0]
	puts "#{expr}"
	(1..ARGV.size-1).each do |i|
		grep(expr, ARGV[i])
	end
end

File: python/day2/grep.py
#!/usr/bin/env python

# A simple grep

# Use python 3 compatible syntax or print
from __future__ import print_function

import sys
import os
import re

# Search for 'expr' in 'file', if no file is provided, stdin is used instead
def grep(expr, file=''):
    try:
        if not file or file == '':
            file = 'stdin'
            f = sys.stdin
        else: 
            f = open(file, "rU")
        for line in f:
            line = line.rstrip('\n')
            if re.search(expr, line):
                print('{}: {}'.format(file, line))
        f.close()
    except IOError:
        print('Error opening file \'{}\'.'.format(file), file=sys.stderr) 
        return       

# Very simple command line parsing: sys.argv[0] == script name, sys.argv[1:] == arguments
if len(sys.argv) < 2:
    # No arguments: print usage message
    script = os.path.basename(sys.argv[0])
    print('Usage: {} expr file...'.format(script), file=sys.stderr) 
else:
    expr =  sys.argv[1]
    print('search expression: {}'.format(expr))
    if len(sys.argv) == 2:
        grep(expr)
    else: 
        for file in sys.argv[2:]:
            grep(expr, file)
